Only prevent full screen video HDR external constraining on macOS

squelart opened the pull request:
#### 1c9259f43c89b497018b0c9503717b3ba56dac54
<pre>
Only prevent full screen video HDR external constraining on macOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=292904">https://bugs.webkit.org/show_bug.cgi?id=292904</a>
<a href="https://rdar.apple.com/151160054">rdar://151160054</a>

Reviewed by Simon Fraser.

In some situations the OS may request that the browser constrains (or
&quot;suppresses&quot;) HDR to some maximum level.
For full screen videos on macOS, we do NOT want to suppress HDR.
But on other OSes like iOS, the constraining request should now be
respected.

Note: This does not affect `dynamic-range-limit`, which is always
respected on all supported platforms.

* LayoutTests/media/video-suppress-hdr-fullscreen-expected.txt: Added.
* LayoutTests/media/video-suppress-hdr-fullscreen.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::computePlayerDynamicRangeLimit const):
* Source/WebCore/page/Settings.yaml:

Canonical link: <a href="https://commits.webkit.org/295178@main">https://commits.webkit.org/295178@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb0fa0cea733f90e136f76a0452201d0ef473690

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104241 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23945 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14288 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109441 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54909 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106281 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24313 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32490 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79169 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/forms/constraints/infinite_backtracking.html (failure)") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107247 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18890 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94049 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59497 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18700 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12091 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54269 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88399 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12150 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111823 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31397 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23142 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88189 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31762 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90242 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87851 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22381 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32744 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10494 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25861 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31326 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36639 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31120 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34456 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32680 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -7591,7 +7591,7 @@ void HTMLMediaElement::didStopBeingFullscreenElement()
 }
 
-#if ENABLE(FULLSCREEN_API)
+#if ENABLE(FULLSCREEN_API) && PLATFORM(MAC)
```
Rather than using PLATFORM(MAC) directly, please use a more specific macro define. Something where the name explains why the condition is needed. (e.g. SUPPORTS(FOO) or ENABLE(CONSTAINT_HDR_IN...))

squelart replied:
Thank you for the suggestion, much more readable now!

annevk commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -1833,6 +1833,10 @@
 #endif
 
+#if PLATFORM(MAC)
+#define ENABLE_FULLSCREEN_VIDEO_PREVENT_HDR_SUPPRESSION 1
```
`ENABLE_` should go in `PlatformEnable.h`. Here it would be something like:

`HAVE_FULLSCREEN_VIDEO_PREVENTS_HDR_SUPPRESSION`

annevk commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -8009,13 +8009,15 @@ PlatformDynamicRangeLimit HTMLMediaElement::computePlayerDynamicRangeLimit() con
             return false;
-#endif
+#endif // ENABLE(FULLSCREEN_API)
+#endif // ENABLE(FULLSCREEN_VIDEO_PREVENT_HDR_SUPPRESSION)
```
These `#endif` comments are the wrong way around.

squelart replied:
They look correct to me:
```
#if ENABLE(FULLSCREEN_VIDEO_PREVENT_HDR_SUPPRESSION)
...
#if ENABLE(FULLSCREEN_API)
...
#endif // ENABLE(FULLSCREEN_API)
#endif // ENABLE(FULLSCREEN_VIDEO_PREVENT_HDR_SUPPRESSION)
```

annevk commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -7607,7 +7607,7 @@ void HTMLMediaElement::didStopBeingFullscreenElement()
 }
 
-#if ENABLE(FULLSCREEN_API)
+#if ENABLE(FULLSCREEN_API) && ENABLE(FULLSCREEN_VIDEO_PREVENT_HDR_SUPPRESSION)
```
This doesn't seem right, but maybe it is. Why should only platforms with `ENABLE(FULLSCREEN_VIDEO_PREVENT_HDR_SUPPRESSION)` have `documentFullscreenChanged`? At least `m_isChildOfElementFullscreen` doesn't seem very HDR-specific.

squelart replied:
It's because passing along the full screen state is only needed in situations where it _may_ influence the suppress-HDR handling. So on non-Mac platforms, it would take unnecessary work and space to pass and store this value, as it would never be read.
I was hoping that having both `FULLSCREEN` and `HDR` in the macro name would convey that special connection here.
Now, this event should be fairly rare (only when full screen state changes), and the storage is only 1 byte, so it may be okay not to optimize this. Other opinions welcome.

annevk replied:
That makes sense, just wanted to double check. Looks good to me with the other comments addressed.

squelart committed:
Only prevent full screen video HDR external constraining on macOS
https://bugs.webkit.org/show_bug.cgi?id=292904
rdar://151160054

Reviewed by Simon Fraser.

In some situations the OS may request that the browser constrains (or
"suppresses") HDR to some maximum level.
For full screen videos on macOS, we do NOT want to suppress HDR.
But on other OSes like iOS, the constraining request should now be
respected.

Note: This does not affect `dynamic-range-limit`, which is always
respected on all supported platforms.

* LayoutTests/media/video-suppress-hdr-fullscreen-expected.txt: Added.
* LayoutTests/media/video-suppress-hdr-fullscreen.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::computePlayerDynamicRangeLimit const):
* Source/WebCore/page/Settings.yaml:

Canonical link: https://commits.webkit.org/295178@main
